src/webgpu/api/validation/encoding/programmable/pipeline_bind_group_compat.spec.ts

```diff
@@ -14,8 +14,301 @@ TODO: subsume existing test, rewrite fixture as needed.
+      textureType = 'texture_storage_2d';
+    } else {
+      switch (binding.storageTexture.viewDimension) {
+        case '1d': {
+          textureType = 'texture_storage_1d';
+          break;
+        }
+        case '2d': {
+          textureType = 'texture_storage_2d';
+          break;
+        }
+        case '2d-array':
+        case '3d': {
+          textureType = 'texture_storage_3d';
+          break;
+        }
+        default:
+          // Cannot be 'cube' or 'cube-array'
+          unreachable();
+      }
+    }
+
+    switch (binding.storageTexture.format) {
+      // float
+      case 'rgba8unorm':
+      case 'rgba8snorm':
+      case 'rgba16float':
+      case 'r32float':
+      case 'rg32float':
+      case 'rgba32float': {
+        texelFormat = 'f32';
+        break;
+      }
+
+      // uint
+      case 'rgba8uint':
+      case 'rgba16uint':
+      case 'r32uint':
+      case 'rg32uint':
+      case 'rgba32uint': {
+        texelFormat = 'u32';
+        break;
+      }
+
+      // sint
+      case 'rgba8sint':
+      case 'rgba16sint':
+      case 'r32sint':
+      case 'rg32sint':
+      case 'rgba32sint': {
+        texelFormat = 'i32';
+        break;
+      }
+      default:
+        unreachable();
+    }
+
+    suffix = `${textureType}<${texelFormat}`;
+
+    if (binding.storageTexture.access === 'write-only') {
+      suffix += ', write>';
+    } else {
+      suffix += '>';
+    }
+
+    declare = `var storeTex: ${suffix}`;
+  } else if (binding.texture !== undefined) {
+    let textureType: string;
+    let texelFormat: string = '';
+
+    const isMultisampled =
+      binding.texture.multisampled !== undefined && binding.texture.multisampled;
+    assert(
+      isMultisampled &&
+        (binding.texture.viewDimension === undefined || binding.texture.viewDimension === '2d')
+    );
+    let isDepth: boolean = false;
+
+    if (binding.texture.sampleType === undefined) {
+      texelFormat = 'f32';
+    } else {
+      switch (binding.texture.sampleType) {
+        case 'float':
+        case 'unfilterable-float': {
+          assert(!isMultisampled);
+          texelFormat = '<f32>';
+          break;
+        }
+        case 'sint': {
+          texelFormat = '<i32>';
+          break;
+        }
+        case 'uint': {
+          texelFormat = '<u32>';
+          break;
+        }
+        case 'depth': {
+          isDepth = true;
+          break;
+        }
+        default:
+          unreachable();
+      }
+    }
+
+    if (binding.texture.viewDimension === undefined) {
+      if (isDepth && isMultisampled) {
+        textureType = 'texture_depth_multisampled_2d';
+      } else if (isMultisampled) {
+        textureType = 'texture_multisampled_2d';
+      } else if (isDepth) {
+        textureType = 'texture_depth_2d';
+      } else {
+        textureType = 'texture_2d';
+      }
+    } else {
+      switch (binding.texture.viewDimension) {
```

Comment:
should also use `??` instead of separate if; specifying `"2d"` should be the same as undefined since "2d" is the default:
https://gpuweb.github.io/gpuweb/#dictdef-gputexturebindinglayout